### PR TITLE
Fixing bug 1710 and adding unit tests

### PR DESCRIPTION
**Pull request by dsgouda**
@olydis PTAL
https://github.com/Azure/autorest/issues/1710

### CODE REVIEW

File: src/modeler/AutoRest.Swagger/OperationBuilder.cs [hunk 1]
```diff
@@ -249,6 +249,10 @@ private List<Stack<IModelType>> BuildResponses(Method method, CompositeType head
                             methodName,
                             response.Key));
                     }
+                    foreach (var key in _operation.Responses[response.Key].Extensions.Keys)
```
Comment (olydis) [inline src/modeler/AutoRest.Swagger/OperationBuilder.cs]: I'd rather copy `_operation.Responses[response.Key].Extensions` to `method.Responses[response.Key.ToHttpStatusCode()].Extensions` directly (so yes, make the property settable, it's mutable anyways).

**Comment by dsgouda** [comment]
@olydis the code should consider every response type and check if it has set x-nullable to false and only then remove the nullability property (this does not take into consideration the "default" response which is generally the error object) I have modified the test yaml to reflect these cases just to be sure

**Comment by dsgouda** [comment]
@azuresdkci Test this please

File: src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml [hunk 2]
```diff
@@ -353,12 +353,12 @@ else
         @:{
             if (Model.DefaultResponse.Body.IsPrimaryType(KnownPrimaryType.Stream))
             {
-            @:@(Model.DefaultResponse.Body.AsNullableType()) _errorBody = await _httpResponse.Content.ReadAsStreamAsync().ConfigureAwait(false);
+            @:@(Model.DefaultResponse.Body.AsNullableType(Model.IsXNullableReturnType)) _errorBody = await _httpResponse.Content.ReadAsStreamAsync().ConfigureAwait(false);
```
Comment (olydis) [inline src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml]: Careful: this seems to be the error body, which is orthogonal to the "return type" body! So this does not belong here.

File: src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml [hunk 3]
```diff
@@ -353,12 +353,12 @@ else
...
             }
             else
             {
             @:_responseContent = await _httpResponse.Content.ReadAsStringAsync().ConfigureAwait(false);
-            @:@(Model.DefaultResponse.Body.AsNullableType()) _errorBody =  Microsoft.Rest.Serialization.SafeJsonConvert.DeserializeObject<@(Model.DefaultResponse.Body.AsNullableType())>(_responseContent, @(Model.GetDeserializationSettingsReference(Model.DefaultResponse.Body)));
+            @:@(Model.DefaultResponse.Body.AsNullableType()) _errorBody =  Microsoft.Rest.Serialization.SafeJsonConvert.DeserializeObject<@(Model.DefaultResponse.Body.AsNullableType(Model.IsXNullableReturnType))>(_responseContent, @(Model.GetDeserializationSettingsReference(Model.DefaultResponse.Body)));
```
Comment (olydis) [inline src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml]: same as above

File: src/generator/AutoRest.CSharp/Model/MethodCs.cs [hunk 4]
```diff
@@ -261,11 +261,11 @@ public virtual string ReturnTypeString
             {
                 if (ReturnType.Body != null)
                 {
-                    return ReturnType.Body.AsNullableType(HttpMethod != HttpMethod.Head);
+                    return ReturnType.Body.AsNullableType(HttpMethod != HttpMethod.Head && IsXNullableReturnType);
                 }
                 if (ReturnType.Headers != null)
                 {
-                    return ReturnType.Headers.AsNullableType(HttpMethod != HttpMethod.Head);
+                    return ReturnType.Headers.AsNullableType(HttpMethod != HttpMethod.Head && IsXNullableReturnType);
```
Comment (olydis) [inline src/generator/AutoRest.CSharp/Model/MethodCs.cs]: nullability does not apply to headers type (that is always an object)

File: src/generator/AutoRest.CSharp/Model/MethodCs.cs [hunk 5]
```diff
@@ -167,12 +167,12 @@ public virtual string TaskExtensionReturnTypeString
...
                 {
                     return string.Format(CultureInfo.InvariantCulture,
-                        "System.Threading.Tasks.Task<{0}>", ReturnType.Body.AsNullableType(HttpMethod != HttpMethod.Head));
+                        "System.Threading.Tasks.Task<{0}>", ReturnType.Body.AsNullableType(HttpMethod != HttpMethod.Head && IsXNullableReturnType));
                 }
                 else if (ReturnType.Headers != null)
                 {
                     return string.Format(CultureInfo.InvariantCulture,
-                        "System.Threading.Tasks.Task<{0}>", ReturnType.Headers.AsNullableType(HttpMethod != HttpMethod.Head));
+                        "System.Threading.Tasks.Task<{0}>", ReturnType.Headers.AsNullableType(HttpMethod != HttpMethod.Head && IsXNullableReturnType));
```
Comment (olydis) [inline src/generator/AutoRest.CSharp/Model/MethodCs.cs]: nullability does not apply to headers type (that is always an object)

File: src/generator/AutoRest.CSharp/Model/MethodCs.cs [hunk 6]
```diff
@@ -141,15 +141,15 @@ public virtual string OperationResponseReturnTypeString
...
                 }
                 if (ReturnType.Body != null)
                 {
-                    return $"Microsoft.Rest.HttpOperationResponse<{ReturnType.Body.AsNullableType(HttpMethod != HttpMethod.Head)}>";
+                    return $"Microsoft.Rest.HttpOperationResponse<{ReturnType.Body.AsNullableType(HttpMethod != HttpMethod.Head && IsXNullableReturnType)}>";
                 }
                 if (ReturnType.Headers != null)
                 {
-                    return $"Microsoft.Rest.HttpOperationHeaderResponse<{ReturnType.Headers.AsNullableType(HttpMethod != HttpMethod.Head)}>";
+                    return $"Microsoft.Rest.HttpOperationHeaderResponse<{ReturnType.Headers.AsNullableType(HttpMethod != HttpMethod.Head && IsXNullableReturnType)}>";
```
Comment (olydis) [inline src/generator/AutoRest.CSharp/Model/MethodCs.cs]: nullability does not apply to headers type (that is always an object)

File: src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml (hunk 2 above)
Comment (dsgouda) [inline src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml]: Yup, that's a good catch

File: src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml (hunk 3 above)
Comment (olydis) [inline src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml]: @dsgouda still wrong

File: src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml [hunk 7]
```diff
@@ -479,7 +479,7 @@ else
         }
         else
         {
-        @:_result.Body = Microsoft.Rest.Serialization.SafeJsonConvert.DeserializeObject<@(Model.DefaultResponse.Body.AsNullableType())>(_defaultResponseContent, @(Model.GetDeserializationSettingsReference(Model.DefaultResponse.Body)));
+        @:_result.Body = Microsoft.Rest.Serialization.SafeJsonConvert.DeserializeObject<@(Model.DefaultResponse.Body.AsNullableType(Model. IsXNullableReturnType))>(_defaultResponseContent, @(Model.GetDeserializationSettingsReference(Model.DefaultResponse.Body)));
```
Comment (olydis) [inline src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml]: also not affected by nullability

Comment (dsgouda) [inline src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml]: What if the default response is just a non-nullable primary type?

File: src/core/AutoRest.Core/Model/Method.cs [hunk 8]
```diff
@@ -268,5 +268,8 @@ public bool StructurallyEquals(Method other)
         [JsonIgnore]
         public virtual HashSet<string> LocallyUsedNames { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
 
+        [JsonIgnore]
+        public virtual bool IsXNullableReturnType { get { return (Responses.Keys.Any(key => Responses[key].Extensions?.Get<bool>("x-nullable") ?? true)); } }
```
Comment (olydis) [inline src/core/AutoRest.Core/Model/Method.cs]: In https://github.com/Azure/autorest/pull/1804#issuecomment-278780237 it sounds like you check if *all* responses response types are marked as nullable, but this checks whether *one* response is. That's why I asked for more tests, we would almost have missed this. Please don't only add tests that test for the "good" behavior.

File: src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml (hunk 7 above)
Comment (olydis) [inline src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml]: valid point, but does `Model.IsXNullableReturnType` reflect whether the default response is supposed to be nullable? thought it's about the succeeding responses.

Comment (dsgouda) [inline src/generator/AutoRest.CSharp/Templates/Rest/Client/MethodTemplate.cshtml]: On second thoughts, seems like you might be right, we do not consider ixnullable for default response types

**Pull request merged**